src/libraries/Microsoft.Extensions.Logging.Console/src/WindowsLogConsole.cs

```diff
@@ -17,42 +17,108 @@ public WindowsLogConsole(bool stdErr = false)
                 ResetColor();
             }
         }
 
-        public void WriteLine(string message, ConsoleColor? background, ConsoleColor? foreground)
+        public void Write(string message)
         {
-            bool colorChanged = SetColor(background, foreground);
-            _textWriter.WriteLine(message);
-            if (colorChanged)
+            // writes content with the embedded foreground and background ansi colored. it skips non-color related ansi codes
+            (int startIndex, int length, ConsoleColor? bg, ConsoleColor? fg) content = (-1, 0, null, null);
+            var span = message.AsSpan();
+            const char EscapeChar = '\x1B';
+            ConsoleColor? color = null;
+            bool isDarkColor = true;
+            for (int i = 0; i < span.Length; i++)
             {
-                ResetColor();
+                if (span[i] != EscapeChar || span.Length < i + 3 || span[i + 1] != '[')
+                {
+                    // not an escape char
+                    content.length++;
+                    if (content.startIndex == -1)
+                    {
+                        content.startIndex = i;
+                    }
+                }
+                else if (span[i + 3] == 'm')
+                {
+                    if (int.TryParse(span.Slice(i + 2, length: 1), out int escapeCode))
+                    {
+                        if (escapeCode == 1)
+                            isDarkColor = false;
+                        // parsing only ansi color codes
+                        i += 3;
+                    }
+                }
+                else if (span[i + 4] == 'm')
+                {
+                    if (int.TryParse(span.Slice(i + 2, length: 2), out int escapeCode))
+                    {
+                        if (SetsForegroundColor(escapeCode, isDarkColor, out color))
+                        {
+                            if (content.startIndex != -1)
+                            {
+                                Write(span.Slice(content.startIndex, content.length), content.bg, content.fg);
+                                content.startIndex = -1;
+                                content.length = 0;
+                                isDarkColor = true;
+                            }
+                            content.fg = color;
+                        }
+                        else if (SetsBackgroundColor(escapeCode, out color))
+                        {
+                            if (content.startIndex != -1)
+                            {
+                                Write(span.Slice(content.startIndex, content.length), content.bg, content.fg);
+                                content.startIndex = -1;
+                                content.length = 0;
+                            }
+                            content.bg = color;
+                        }
+                        // parsing only ansi color codes
+                        i += 4;
```

Comment:
>4 [](start = 29, length = 1)

should be +5?

Reply:
never mind we increment it in the loop.


---
In reply to: [451227810](https://github.com/dotnet/runtime/pull/38616#discussion_r451227810) [](ancestors = 451227810)